File: Frontend/src/app/core/services/VacancyService.ts
import { Injectable } from '@angular/core';
import { ApiPathProvider } from './../configs/apiPathProvider';
import { Inject } from '@angular/core';
import { AliasIHttpProvider, IHttpProvider } from '../interfaces/IHttpProvider';
import { RequestOptions, Headers } from '@angular/http';
import { Vacancy } from '../models/vacancy';

@Injectable()
export class VacancyService {

    basePath: string;
    constructor(@Inject(AliasIHttpProvider)private httpProvider: IHttpProvider, apiPathConfiguration: ApiPathProvider) {
        this.httpProvider.basePath = apiPathConfiguration.basePath;
    }

    GetAll(page: number, pageSize: number) {
        const requestOptions = new RequestOptions();
        const headers = new Headers();
        requestOptions.headers = headers;
        return this.httpProvider.Get('Vacancies?page=' + page + '&pageSize=' + pageSize, requestOptions);
    }

    GetById(id: string) {
        const requestOptions = new RequestOptions();
        const headers = new Headers();
        requestOptions.headers = headers;
        return this.httpProvider.Get('Vacancies/' + id, requestOptions);
    }

    Create(item: Vacancy) {
        const requestOptions = new RequestOptions();
        const headers = new Headers();
        requestOptions.headers = headers;

        const body = new FormData();
        body.append('Title', item.title);
        body.append('WorkingCondition', item.workingCondition.toString());
        body.append('Description', item.description);
        body.append('City', item.city);
        body.append('IsRemoteWorking', item.isRemoteWorking.toString());
        body.append('OrganizationName', item.organizationName);
        body.append('Email', item.email);
        body.append('CompanyURL', item.companyURL);
        body.append('PhoneNumber', item.phoneNumber);
        body.append('ContactPartner', item.contactPartner);
        body.append('Salary', item.salary.toString());
        body.append('CurrencyType', item.currencyType.toString());

        return this.httpProvider.Post('Vacancies/', body, requestOptions);
    }

/*
    Post(path: string, body: any, options: RequestOptions) {
        return this.http.post(this.basePath + path, body, options);
    }*/

    /*Put(path: string, body: any, options: RequestOptions) {
        return this.http.put(this.basePath + path, body, options);
    }

    Delete(path: string, options: RequestOptions) {
        return this.http.delete(this.basePath + path, options);
    }*/
}

File: Backend/testTask/Services/VacancyService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using testTask.Configure;
using testTask.Contexts;
using testTask.Entities;
using testTask.Enums;
using testTask.Interfaces;

namespace testTask.Services
{
    public class VacancyService : IVacancyService
    {
        private readonly ISocialWebService _socialWebService;
        private readonly ILogger<VacancyService> _logger;
        private readonly AppDbContext _appDbContext;
        private readonly IOptions<AppOptions> _appOptions;
        public VacancyService(ILogger<VacancyService> logger, AppDbContext appDbContext, 
            ISocialWebService socialWebService, IOptions<AppOptions> appOptions)
        {
            _appOptions = appOptions;
            _socialWebService = socialWebService;
            _logger = logger;
            _appDbContext = appDbContext;
        }

        public async Task Create(string title, WorkingConditions workingCondition, string description, string city, 
            bool isRemoteWorking, string organizationName, string email, string companyURL, string phoneNumber, 
            string contactPartner, decimal salary, CurrencyTypes currencyType)
        {
            try
            {
                var item = new Vacancy
                {
                    Title = title,
                    WorkingCondition = workingCondition,
                    Description = description,
                    City = city,
                    IsRemoteWorking = isRemoteWorking,
                    OrganizationName = organizationName,
                    Email = email,
                    CompanyURL = companyURL,
                    PhoneNumber = phoneNumber,
                    ContactPartner = contactPartner,
                    Salary = salary,
                    CurrencyType = currencyType,
                    Date = DateTime.Now
                };
                await _appDbContext.AddAsync(item);
                await _appDbContext.SaveChangesAsync();
                var message = string.Format("{0} @ {1} #job #.net {2}/{3}/{4}", 
                    item.Title, item.OrganizationName, _appOptions.Value.FrontDNS, _appOptions.Value.VacancyDetailPath, item.Id);
                await _socialWebService.SendInformationAsync(message);
            }
            catch (Exception e)
            {
                _logger.LogError(e.Message + e.StackTrace);
                throw new Exception();
            }
        }

        public async Task<ICollection<Vacancy>> Get(int page, int pageSize)
        {
            var result = new List<Vacancy>();
            try
            {
                result = await _appDbContext.Vacancies.Skip((page - 1) * pageSize).Take(pageSize).ToListAsync();
            }
            catch(Exception e)
            {
                _logger.LogError(e.Message + e.StackTrace);
                throw new Exception();
            }
            return result;
        }

        public Task<Vacancy> Get(string id)
        {
            return _appDbContext.Vacancies.FirstOrDefaultAsync(x => x.Id == id);
        }

        public async Task<int> GetPageCount(int pageSize)
        {
            var count = await _appDbContext.Vacancies.CountAsync();
            var result = count / pageSize;
            if (count % pageSize > 0) result++;
            return result;
        }
    }
}
